Improve messaging of ProcessGroupNCCL destructor

kwen2501 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #142297
* #141511
* #141510

And removed some unnecessary conditions for calling `thread.join()` -- `thread.joinable()` should have covered it.

cc @H-Huang @awgu @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @c-p-i-o

kwen2501 committed:
Improve messaging of ProcessGroupNCCL destructor

[ghstack-poisoned]

wconstab commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -1472,44 +1472,52 @@ void ProcessGroupNCCL::shutdown() {
-    if (ncclHeartbeatMonitorThread_.joinable()) {
-      ncclHeartbeatMonitorThread_.join();
-      VLOG(2) << logPrefix()
+  if (ncclCommWatchdogThread_.joinable()) {
```
whats the reasoning for removing the conditional on `!blockingWait_`?

kwen2501 replied:
If blockingWait_ is true, we would not have spawned the side threads, and `joinable()` would return false.

wconstab replied:
hm. for all the threads?

kwen2501 replied:
Yeah, this is a general thing. If a `std::thread` is just declared but not spawned, it should not be `joinable()`.

wconstab commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -1472,44 +1472,52 @@ void ProcessGroupNCCL::shutdown() {
-    // Here we route to `abort()` explicitly to maintain the old behavior, until
-    // we fix everything.
-    abort();
+  if (terminateProcessGroup_.load())
```
iiuc this is supposed to be the same logic as above? (no behavior change?)

kwen2501 replied:
Right, no behavior change. Just removing condition cascading.

wconstab commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -1472,44 +1472,52 @@ void ProcessGroupNCCL::shutdown() {
 
+  // Note: we have rewritten `shutdown` to represent the destroy behavior
+  // (blocking). Thus we route to `abort()` explicitly here to maintain the
+  // old behavior. We should re-validate the choice and document the rationale
```
can you make this comment more explicit?  ("the old behavior", "have rewritten shutdown" both a bit vague)

what is the TODO part, do we want to remove this abort or do we want to switch this abort to a shutdown? and why..

kwen2501 committed:
Update on "Improve messaging of ProcessGroupNCCL destructor"


And removed some unnecessary conditions for calling `thread.join()` -- `thread.joinable()` should have covered it.

cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

kwen2501 commented:
@pytorchbot merge -f "CI was green; just changing comments"
